File: algo/146. LRU Cache.js
// @9 hashmap
// always remember to set both prev and next while using double linked list

function Node(key, val) {
    this.key = key;
    this.val = val;
    this.next = null;
    this.prev = null;
}

/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.capacity = capacity;
    this.helper_dict = {};
    this.head = null;
    this.tail = null;
    this.currSize = 0;
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if (this.helper_dict.hasOwnProperty(key)) {
        if (this.helper_dict[key] === this.tail) {
            return this.helper_dict[key].val;
        }
        var temp = this.helper_dict[key].prev;
        if (temp !== null) {
            temp.next = temp.next.next;
            temp.next.prev = temp;
        } else {
            this.head = this.helper_dict[key].next;
            this.head.prev = null;
        }
        this.tail.next = this.helper_dict[key];
        this.helper_dict[key].prev = this.tail;
        this.tail = this.helper_dict[key];
        this.tail.next = null;
        return this.helper_dict[key].val;
    }
    return -1;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    //console.log(this.helper_dict);
    //console.log(this.currSize);
    if (this.helper_dict.hasOwnProperty(key)) {
        this.helper_dict[key].val = value;
        this.get(key);
        return;
    } else {
        this.currSize++;
        var node = new Node(key, value);
        if (this.head == null && this.tail == null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            node.prev = this.tail;
            this.tail = node;
        }
        this.helper_dict[key] = node;
        //console.log('currSize', this.currSize, 'capacity', this.capacity, 'this key', key, 'head key', this.head.key);
    }
    if (this.currSize > this.capacity) {
        var tempKey = this.head.key;
        this.head = this.head.next;
        this.head.prev = null;
        delete this.helper_dict[tempKey];        
        this.currSize--;
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = Object.create(LRUCache).createNew(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */

File: round2/146. LRU Cache.cs
public class Node {
    public int key {get; set;}
    public int val {get; set;}
    public Node next {get; set;}
    public Node prev {get; set;}
    public Node(int k, int v) {
        key = k;
        val = v;
        next = null;
        prev = null;
    }
}

public class LRUCache {
    private int _capacity;
    private int Count;
    private Dictionary<int, Node> m;
    private Node head;
    private Node tail;

    public LRUCache(int capacity) {
        _capacity = capacity;
        Count = 0;
        m = new Dictionary<int, Node>();
        head = null;
        tail = null;
    }
    
    public int Get(int key) {
        if (!m.ContainsKey(key)) {
            return -1;
        }
        if (tail == m[key]) return tail.val;
        Node target = m[key];
        if (target.prev != null) {
            target.prev.next = target.next;
            target.next.prev = target.prev;
        } else {
            head = target.next;
            target.next.prev = null;
        }
        target.next = null;
        target.prev = tail;
        tail.next = target;
        tail = target;
        return tail.val;
    }
    
    public void Put(int key, int value) {
        if (m.ContainsKey(key)) {
            m[key].val = value;
            Get(key);
        } else {
            var node = new Node(key, value);
            m[key] = node;
            if (Count == 0) {               
                head = node;
                tail = node;
                Count++;
                return;
            } else {
                tail.next = node;
                node.prev = tail;
                tail = node;
                if (Count < _capacity) {
                    Count++;
                    return;
                } else {
                    m.Remove(head.key);
                    head = head.next;
                    head.prev = null;
                }
            }
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */